### ISSUE TITLE
Concurrency verification

### ISSUE THREAD

**Issue by sebastienros** [OrchardCMS/OrchardCore#1813]
Based on some work that has been done in O1, we need to ensure that we can handle most concurrent patterns with with unit tests or functional tests:
- Creating content items concurrently from APIs (Content Manager) and from the UI
- Should be tested on all parts as they might expose specific data access issues: Autoroute, Menu, ...
- Tested on SQL Server at least (ideally all databases)

**Comment by ArashSade** [comment]
Indeed. Very important for every multi-editor scenario. Strange that this is missing!
Because it is one of those features that if you do not build from the beginning, it would be very difficult to add afterwards.

**Comment by blueheron786** [comment]
Looks like there's a fair amount that's unit testable, although most of it requires a `SiteContext` and underlying DB. I'll take a stab at the unit tests first, and then maybe another look at integratino tests after.

**Comment by blueheron786** [comment]
Not sure what happened to my previous comment. This looks really complicated and is probably best left to someone with more experience than the code-base than I have.

It looks like some classes use `ISession` but internally call `_session.QueryIndex` and other extension methods. Since they're just static classes with syntactic sugar, it's impossible to mock them; we need to use a real DB. Like SQLite, although that might not be a good choice for concurrency tests, since it limits reads or writes or something (can't remember the details).

For whoever works on this next, and for whatever it's worth, I recommend going for integration tests instead. It's cleaner, and gives a better "overall" test (real DB + real code together) for multiple tasks executing at once.

**Comment by Piedone** [comment]
While I agree that this should rather be approached with integration tests, FYI `ISession` can be fully mocked. The static extension methods just chain and parameterize mockable methods too.

**Comment by blueheron786** [comment]
Looks like you're right. Also looks like there's opportunity to add non-concurrency-related unit tests for logic, too. Will open a PR.
